src/util/irep.h

```diff
@@ -242,26 +248,175 @@ class irept
 
   // Note that the move assignment operator does avoid
   // three branches compared to standard operator above.
-  irept &operator=(irept &&irep)
+  sharing_treet &operator=(sharing_treet &&irep)
   {
-    #ifdef IREP_DEBUG
+#ifdef IREP_DEBUG
     std::cout << "ASSIGN MOVE\n";
-    #endif
+#endif
     // we simply swap two pointers
     std::swap(data, irep.data);
     return *this;
   }
 
-  ~irept()
+  ~sharing_treet()
   {
     remove_ref(data);
   }
 
-  #else
-  irept()
+protected:
+  dt *data;
+  static dt empty_d;
+
+  static void remove_ref(dt *old_data);
+  static void nonrecursive_destructor(dt *old_data);
+  void detach();
+
+public:
+  const dt &read() const
   {
+    return *data;
   }
-  #endif
+
+  dt &write()
+  {
+    detach();
+#ifdef HASH_CODE
+    data->hash_code = 0;
+#endif
+    return *data;
+  }
+};
+
+// Static field initialization
+template <typename derivedt>
+typename sharing_treet<derivedt>::dt sharing_treet<derivedt>::empty_d;
+
+/// Base class for tree-like data structures without sharing
+template <typename derivedt>
+class non_sharing_treet
+{
+public:
+  using dt = tree_nodet<derivedt, false>;
+  using subt = typename dt::subt;
+  using named_subt = typename dt::named_subt;
+
+  /// Used to refer to this class from derived classes
+  using tree_implementationt = non_sharing_treet;
+
+  explicit non_sharing_treet(const irep_idt &_id)
+  {
+    write().data = _id;
+  }
+
+  non_sharing_treet(
```

Comment:
Why does `non_sharing_treet` have different constructors? Surely the same take-by-value-and-move concept applies here as does for `sharing_treet`?